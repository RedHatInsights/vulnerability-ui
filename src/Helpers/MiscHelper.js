import { SecurityIcon, UnknownIcon } from '@patternfly/react-icons';
import { SortByDirection } from '@patternfly/react-table';
import findIndex from 'lodash/findIndex';
import propTypes from 'prop-types';
import React from 'react';
import { impactColorList, impactList, PUBLIC_DATE_OPTIONS } from './constants';
import qs from 'query-string';
import { IntlProvider } from '@redhat-cloud-services/frontend-components-translations';

export const dataShape = propTypes.shape({
    data: propTypes.oneOfType([propTypes.object, propTypes.array]),
    meta: propTypes.object,
    isLoading: propTypes.bool
});

/**
 * Based on the  allowedParams it will construct the arguments for the API call
 * which is defined in the vulnerabilities-client. vulnerabilities-client is written is typescript
 * so we need to define and pass every single parameter. Order in allowedParams is important
 *
 * @param {Object} apiProps
 * @param {Array} allowedParams - order should align with the params in the vulnerabilities-client
 */
export function constructParameters(apiProps, allowedParams) {
    if (apiProps) {
        Object.keys(apiProps).forEach(
            key => (
                apiProps[key] === undefined
                || apiProps[key] === ''
            )
            && delete apiProps[key]
        );

        const params = allowedParams.map(item => apiProps[item]);

        return params;
    }

    return [];
}

// TODO DRY:similar to constructParameters
export function constructURLParameters(urlParams, allowedParams) {
    if (urlParams) {
        const params = { ...urlParams };
        Object.keys(urlParams).forEach(
            key => (
                params[key] === undefined
                || params[key] === ''
                || !allowedParams.includes(key)
                || params[key] === false
            )
            && delete params[key]

        );
        return params;
    }
}

export function formatDate(date = new Date(), includeTime = false) {
    const prepend = (number) => `${`${number}`.length === 1 ? '0' : ''}${number}`;
    const toFormat = new Date(date);
    if (toFormat instanceof Date && !isNaN(toFormat)) {
        const year = toFormat.getFullYear();
        const month = toFormat.getMonth() + 1; // month is zero indexed
        const day = toFormat.getDate();
        const hour = toFormat.getUTCHours();
        const minute = toFormat.getUTCMinutes();

        return includeTime
            ? `${year}-${prepend(month)}-${prepend(day)} ${prepend(hour)}:${prepend(minute)} UTC`
            : `${year}-${prepend(month)}-${prepend(day)}`;
    }

    return toFormat;
}

/**
 * Contructs the filters which have more than one parameter that needs to be passed
 * and resets all the parameteres when it's needed
 *
 * @param {Object} filterParams
 * @return {Object}
 */
export function constructFilterParameters(filterParams) {
    if (Object.prototype.hasOwnProperty.call(filterParams, 'publish_date')) {
        let publicEntry = PUBLIC_DATE_OPTIONS.find(item => item.value === filterParams.publish_date);

        if (typeof publicEntry !== 'undefined') {
            filterParams.public_from = publicEntry.from && formatDate(publicEntry.from);
            filterParams.public_to = publicEntry.to && formatDate(publicEntry.to);
        } else {
            filterParams.public_from = undefined;
            filterParams.public_to = undefined;
            filterParams.publish_date = undefined;
        }
    }

    return filterParams;
}

export function getImpactDetails(impact) {
    if (Object.keys(impactList).includes(impact)) {
        return {
            title: impactList[impact].title,
            color: impactList[impact].color,
            text: impactList[impact].text,
            titleContent: (
                <SecurityIcon size="lg" />
            )
        };
    } else {
        return {
            title: 'Unknown',
            color: impactColorList.unknown,
            text: '',
            titleContent: (
                <UnknownIcon size="lg" />
            )
        };
    }
}

export const createSortBy = (header, value) => {
    if (value) {
        let direction = value.indexOf('-') === 0 ? SortByDirection.desc : SortByDirection.asc;
        value = value.replace(/^(-|\+)/, '');
        const index = findIndex(header, item => item.key === value);
        let sort = {
            index,
            direction
        };
        return sort;
    }

    return {};
};

export const handleSortColumn = (key, direction, header, currentSort, apply) => {
    let columnName = header[key].key;
    const useDefault = currentSort && currentSort.substr(1) !== columnName;
    if (direction === SortByDirection.desc || useDefault) {
        columnName = '-' + columnName;
    }

    apply({ sort: columnName });
};

export const updateStateSet = (stateSet, names, value) => {
    [].concat(names).forEach(name => {
        value ? stateSet.add(name) : stateSet.delete(name);
    });
    return stateSet;
};

export const useUrlParams = (allowedParams) => {
    const url = new URL(window.location);
    const urlParams = qs.parse(url.search);

    const setUrlParams = (parameters) => {
        const searchParams = qs.stringify(constructURLParameters(parameters, allowedParams));
        window.history.replaceState(null, null, `${url.origin}${url.pathname}?${searchParams}`);
    };

    return [urlParams, setUrlParams];
};

export const updateRef = (meta, apply) => {
    const pages = parseInt(meta.pages);
    const page = parseInt(meta.page);
    const cvesCount = parseInt(meta.cvesCount);
    const goTo = (pages === page && page > 1 && cvesCount === 1) ? (pages - 1) : page;

    apply({ page: goTo });
};

export const mountWithIntl = (Component) => {
    const wrapper = mount(Component, {
        wrappingComponent: IntlProvider
    });
    const provider = wrapper.getWrappingComponent();
    provider.setProps({ locale: 'en' });

    return wrapper;
};

export const shallowWithIntl = (Component) => {
    const wrapper = shallow(Component, {
        wrappingComponent: IntlProvider
    });
    const provider = wrapper.getWrappingComponent();
    provider.setProps({ locale: 'en' });

    return wrapper;
};

export const insertIf = (condition, ...elements) => {
    return condition ? elements : [];
};
