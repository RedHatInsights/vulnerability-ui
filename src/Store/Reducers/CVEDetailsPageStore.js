import Immutable from 'seamless-immutable';
import * as ActionTypes from '../ActionTypes';

// Initial State
export const initialState = Immutable({
    affectedSystemsByCVE: {
        isLoading: true,
        openedRows: [],
        payload: {
            meta: {
                total_items: 0,
                page: 1,
                page_size: 20
            }
        }
    },
    parameters: {
        page: 1,
        page_size: 20,
        selectedHosts: [],
        status_id: undefined,
        sort: '-last_upload',
        show_advisories: true
    },
    cveDetails: {
        isLoading: true,
        payload: {},
        error: false
    }
});

// Reducer
export const CVEDetailsPageStore = (state = initialState, action) => {
    let newState;
    switch (action.type) {
        case ActionTypes.FETCH_CVE_DETAILS + '_REJECTED':
            newState = state.setIn(['cveDetails', 'error'], true);
            newState = newState.setIn(['cveDetails', 'isLoading'], false);
            return newState;
        case ActionTypes.FETCH_CVE_DETAILS + '_PENDING':
            newState = state.setIn(['cveDetails', 'isLoading'], true);
            return newState;

        case ActionTypes.FETCH_CVE_DETAILS + '_FULFILLED':
            newState = state.setIn(['cveDetails', 'payload'], action.payload);
            newState = newState.setIn(['cveDetails', 'isLoading'], false);
            return newState;

        case ActionTypes.FETCH_AFFECTED_SYSTEMS_BY_CVE + '_PENDING':
            newState = state.setIn(['affectedSystemsByCVE', 'isLoading'], true);
            return newState;

        case ActionTypes.FETCH_AFFECTED_SYSTEMS_BY_CVE + '_FULFILLED':
            newState = state.setIn(['affectedSystemsByCVE', 'payload'], action.payload);
            newState = newState.setIn(['affectedSystemsByCVE', 'isLoading'], false);
            return newState;
        case ActionTypes.CHANGE_EXPOSED_SYSTEMS_PARAMETERS:
            newState = state.setIn(['parameters'], {
                ...state.parameters,
                ...action.payload,
                page_size: action.payload.page_size || state.parameters.page_size
            });
            return newState;
        case ActionTypes.SET_GLOBAL_FILTER:
            newState = state.setIn(['parameters'], {
                ...state.parameters,
                ...action.payload
            });
            return newState;
        case ActionTypes.CLEAR_CVE_STORE:
            return initialState;
        default:
            return state;
    }
};

