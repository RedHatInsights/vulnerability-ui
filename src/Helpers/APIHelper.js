let BASE_ROUTE = '/api/vulnerability/v1';
import { GitApi } from '@redhat-cloud-services/vulnerabilities-client';
import { SystemProfileApi } from '@redhat-cloud-services/host-inventory-client';
import instance from '../Utilities/interceptors';
import { constructParameters } from './MiscHelper';

let api = new GitApi(undefined, BASE_ROUTE, instance);

const INVENTORY_API_BASE = '/api/inventory/v1';
const systemProfile = new SystemProfileApi(undefined, INVENTORY_API_BASE, instance);

const defaultParams = [
    'filter',
    'limit',
    'offset',
    'page',
    'page_size',
    'sort'
];

export function getCveDetails(synopsis) {
    const result = api.getCveDetails(synopsis);
    return result;
}

export function getSystemDetails(system) {
    const result = api.getSystemDetails(system);
    return result;
}

export function setSystemCveStatus(params) {
    let result = api.setStatus(params);
    return result;
}

export function optOutSystems(systems, optOut) {
    const data = { opt_out: optOut, inventory_id: new Array(...systems) };
    let result = api.setSystemsOptOut(data);
    return result;
}

export function setBusinessRisk(params) {
    let result = api.setCveBusinessRisk(params);
    return result;
}

export function setCveStatus(params) {
    let result = api.setCveStatus(params);
    return result;
}

/**
 * /vulnerabilities/cves
 * /vulnerabilities/cves/ids
 */
let cveParams = [
    'cvss_from',
    'cvss_to',
    'public_from',
    'public_to',
    'impact',
    'data_format',
    'business_risk_id',
    'status_id',
    'rule_presence',
    'tags',
    'sap_sids',
    'sap_system',
    'known_exploit',
    'affecting',
    'rhel_version'
];

export function getCveListByAccount(apiProps) {
    let parameterNames = [
        ...defaultParams,
        ...cveParams,
        'report',
        'advanced_report',
        'ansible',
        'mssql'
    ];
    let parameterArray = constructParameters(apiProps, parameterNames);
    let result = api.getCveList(...parameterArray);
    return result;
}

export function getCveIdsList(apiProps) {
    let parameterNames = [
        ...defaultParams,
        ...cveParams,
        'ansible',
        'mssql'
    ];

    let parameterArray = constructParameters(apiProps, parameterNames);
    let result = api.getCveIdsList(...parameterArray);
    return result;
}

/**
 * /cves/ID/affected_systems
 */
let systemsByCVEparams = [
    'status_id',
    'data_format',
    'uuid',
    'rule_key',
    'rule_presence',
    'rule',
    'tags',
    'sap_sids',
    'sap_system',
    'show_advisories',
    'advisory',
    'rhel_version',
    'first_reported_from',
    'first_reported_to',
    'advisory_available',
    'remediation'
];

export function getAffectedSystemsByCVE({ id, ...apiProps }) {
    let parameterNames = [
        ...defaultParams,
        ...systemsByCVEparams,
        'report',
        'ansible',
        'mssql'
    ];

    let parameterArray = constructParameters(apiProps, parameterNames);
    let result = api.getAffectedSystemsByCve(id, ...parameterArray);
    return result;
}

export function getAffectedSystemsIdsByCve({ id, ...apiProps }) {
    let parameterNames = [
        ...defaultParams,
        ...systemsByCVEparams,
        'ansible',
        'mssql'
    ];

    let parameterArray = constructParameters(apiProps, parameterNames);
    let result = api.getAffectedSystemsIdsByCve(id, ...parameterArray);
    return result;
}

/**
 * /systems/ID/cves
 * /systems/ID/cves/ids
 */
let cveBySystemParams = [
    'cvss_from',
    'cvss_to',
    'public_from',
    'public_to',
    'impact',
    'status_id',
    'data_format',
    'business_risk_id',
    'rule_presence',
    'show_advisories',
    'advisory',
    'rule_key',
    'known_exploit',
    'first_reported_from',
    'first_reported_to',
    'advisory_available',
    'remediation'
];

export function getCveListBySystem(apiProps) {
    const { system } = apiProps;
    let parameterNames = [
        ...defaultParams,
        ...cveBySystemParams,
        'report'
    ];

    if (apiProps && system) {
        Object.keys(apiProps).forEach(key => (apiProps[key] === undefined || apiProps[key] === '') && delete apiProps[key]);
        const params = parameterNames.map(item => apiProps[item]);
        return api.getCveListBySystem(system, ...params).catch(err => {
            if (err && err.status === '404') {
                return { errors: err };
            }

            throw err;
        });
    }
}

export function getCveIdsBySystem(apiProps) {
    const { system } = apiProps;
    let parameterNames = [
        ...defaultParams,
        ...cveBySystemParams
    ];

    if (apiProps && system) {
        Object.keys(apiProps).forEach(key => (apiProps[key] === undefined || apiProps[key] === '') && delete apiProps[key]);
        const params = parameterNames.map(item => apiProps[item]);
        return api.getCveIdsBySystem(system, ...params).catch(err => {
            if (err && err.status === '404') {
                return { errors: err };
            }

            throw err;
        });
    }
}

/**
 * /systems/ids
 * /systems
 */
let systemsParams = [
    'data_format',
    'stale',
    'uuid',
    'tags',
    'sap_sids',
    'sap_system',
    'excluded',
    'rhel_version'
];

export function getSystems(apiProps) {
    let parameterNames = [
        ...defaultParams,
        ...systemsParams,
        'report',
        'ansible',
        'mssql'
    ];
    let parameterArray = constructParameters(apiProps, parameterNames);
    let result = api.getSystemsList(...parameterArray);
    return result;
}

export function getSystemsIds(apiProps) {
    let parameterNames = [
        ...defaultParams,
        ...systemsParams
    ];
    let parameterArray = constructParameters(apiProps, parameterNames);
    let result = api.getSystemsIds(...parameterArray);
    return result;
}

export function getAnnouncement() {
    let result = api.getAnnouncement();
    return result;
}

export function getDashbar(apiProps) {
    let parameterNames = [
        'tags',
        'sap_sids',
        'sap_system',
        'ansible',
        'mssql'
    ];
    let parameterArray = constructParameters(apiProps, parameterNames);
    let result = api.getDashbar(...parameterArray);
    return result;
}

export function getOperatingSystems() {
    return systemProfile.apiSystemProfileGetOperatingSystem();
}

export function fetchPdfReport(template, filters) {
    const CRC_PDF_GENERATE_API = '/api/crc-pdf-generator/v1/generate';
    const url = new URL(CRC_PDF_GENERATE_API,  window.location.origin);

    return fetch(url, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json'
        },

        body: JSON.stringify({
            service: 'vulnerability',
            template,
            params: {
                ...filters
            }
        })
    }, 50000)
        .then((response) => response.blob());
}
