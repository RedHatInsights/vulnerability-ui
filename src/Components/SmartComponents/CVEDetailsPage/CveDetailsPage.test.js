import CveDetailsPage from './CVEDetailsPage';
import { mountWithIntl } from '../../../Helpers/MiscHelper';
import { BrowserRouter as Router } from 'react-router-dom';
import { initialState } from '../../../Store/Reducers/CVEDetailsPageStore';
import configureStore from 'redux-mock-store';
import { Provider, useSelector } from 'react-redux';
import toJson from 'enzyme-to-json';
import Immutable from 'seamless-immutable';
import { shallow } from 'enzyme';
import { act } from "react-dom/test-utils";

jest.mock('../../../Helpers/Hooks', () => ({
    ...jest.requireActual('../../../Helpers/Hooks'),
    useRbac: () => [[true, true, true, true], false]
}));

jest.mock("react-redux", () => ({
    ...jest.requireActual("react-redux"),
    useSelector: jest.fn()
}));

let state = {
    ...initialState,
    cveDetails: {
        isLoading: false,
        error: false,
        payload: Immutable({
            data: {
                attributes: {
                    business_risk: "High",
                    business_risk_id: 3,
                    business_risk_text: null,
                    cvss2_metrics: null,
                    cvss2_score: null,
                    cvss3_metrics: "CVSS:3.1/AV:L/AC:L/PR:L/UI:N/S:C/C:H/I:N/A:N",
                    cvss3_score: "6.500",
                    description: "A new domain bypass transient execution attack",
                    modified_date: "2020-07-20T10:37:20+00:00",
                    public_date: "2020-06-09T17:00:00+00:00",
                    redhat_url: "",
                    rules: [],
                    secondary_url: "",
                    status: "In-Review",
                    status_id: 1,
                    status_text: null,
                    synopsis: "CVE-2020-0543",
                    systems_status_detail: '',
                    systems_status_divergent: 73
                }
            }
        })
    }
}

useSelector.mockImplementation(callback => {
    return callback({ CVEDetailsPageStore: state })
});

const customMiddleWare = store => next => action => {
    next(action);
}
const mockStore = configureStore([customMiddleWare]);
const store = mockStore(initialState);
const props = { match: { params: { cve: 'testCve' } } };
window.insights = {
    loadInventory: (args) => new Promise.resolve({
        inventoryConnector: (args) => {
            const InventoryTable = ({ testProp }) => <div>A mock with '{testProp}' passed!</div>;
            return ({ InventoryTable });
        },
        mergeWithEntities: () => { },
        mergeWithDetail: () => { }
    })
};
describe('CVE details:', () => {
    it('Should render match snapshot', () => {
        const wrapper = shallow(<CveDetailsPage {...props} store={store} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('Should dispatch actions to fetch data', () => {
        mountWithIntl(
            <Provider store={store}>
                <Router><CveDetailsPage {...props} /></Router>
            </Provider>
        );
        expect(store.getActions().some(({ type }) => type === 'FETCH_CVE_DETAILS')).toBeTruthy();
        expect(store.getActions().some(({ type }) => type === 'FETCH_AFFECTED_SYSTEMS_BY_CVE')).toBeFalsy();
    });

    describe('test modals: ', () => {
        const wrapper = mountWithIntl(<Provider store={store}>
            <Router><CveDetailsPage {...props} /></Router>
        </Provider>);

        it('Should set BusinessRiskModal', () => {
            const setBusnessRisk = wrapper.find('BaseDropdown').props().dropdownItems[0].props.onClick;
            act(() => setBusnessRisk([{ business_risk_id: 'testCves' }]));
            wrapper.update();
            expect(wrapper.find('BusinessRiskModal')).toBeTruthy();
        });

        it('Should set StatusModal', () => {
            const setStatusModal = wrapper.find('BaseDropdown').props().dropdownItems[1].props.onClick;
            act(() => setStatusModal([{ business_risk_id: 'testCves' }]));
            wrapper.update();
            const statusModal = wrapper.find('CveStatusModal');
            expect(statusModal).toBeTruthy();
            act(() => statusModal.props().updateRef());
            expect(store.getActions()[4].type).toEqual('FETCH_CVE_DETAILS');
        });
    });

    it('Should display InvalidObject page', () => {
        useSelector.mockImplementation(callback => {
            return callback({
                CVEDetailsPageStore: {
                    ...initialState,
                    cveDetails: {
                        isLoading: false,
                        error: {
                            hasError: true
                        },
                    }
                }
            })
        });

        const wrapper = mountWithIntl(<Provider store={store}>
            <Router><CveDetailsPage {...props} /></Router>
        </Provider>);

        expect(wrapper.find('InvalidObject')).toBeTruthy();
        wrapper.unmount();
    });

    it('Should process filterRulevalues and send as a prop to SystemsExposedPage', () => {
        let newState = { ...state };
        newState.cveDetails.payload = Immutable({
            data: {
                attributes: {
                    business_risk: "High",
                    business_risk_id: 3,
                    business_risk_text: null,
                    cvss2_metrics: null,
                    cvss2_score: null,
                    cvss3_metrics: "CVSS:3.1/AV:L/AC:L/PR:L/UI:N/S:C/C:H/I:N/A:N",
                    cvss3_score: "6.500",
                    description: "A new domain bypass transient execution attack",
                    modified_date: "2020-07-20T10:37:20+00:00",
                    public_date: "2020-06-09T17:00:00+00:00",
                    redhat_url: "",
                    rules: [{
                        "associated_cves": [
                            "CVE-2019-11135"
                        ],
                        "change_risk": 3,
                        "description": "testDescription",
                        "generic": "sd",
                        "kbase_node_id": 4572391,
                        "playbook_count": 1,
                        "reboot_required": false,
                        "rule_id": "testID",
                        "summary": "testSummary"
                    }],
                    secondary_url: "",
                    status: "In-Review",
                    status_id: 1,
                    status_text: null,
                    synopsis: "CVE-2020-0543",
                    systems_status_detail: '',
                    systems_status_divergent: 73
                }
            }
        });
        useSelector.mockImplementation(callback => {
            return callback({ CVEDetailsPageStore: newState });
        });

        const wrapper = mountWithIntl(<Provider store={store}>
            <Router><CveDetailsPage {...props} /></Router>
        </Provider>)
        const props = wrapper.find('SystemsExposedTable').props();
        expect(props.filterRuleValues).toEqual([{ value: 'testID', label: 'testDescription' }]);
    });
});
