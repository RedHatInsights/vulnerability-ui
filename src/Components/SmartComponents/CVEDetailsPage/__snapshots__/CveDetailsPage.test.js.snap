// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CVE details: Should render match snapshot 1`] = `
<DocumentFragment>
  <section
    class="pf-v5-l-page-header pf-v5-c-page-header pf-v5-l-page__main-section pf-v5-c-page__main-section pf-m-light"
    data-ouia-component-id="OUIA-Generated-RHI/Header-true-1"
    data-ouia-component-type="RHI/Header"
    data-ouia-safe="true"
    widget-type="InsightsPageHeader"
  >
    <nav
      aria-label="Breadcrumb"
      class="pf-v5-c-breadcrumb pf-v5-u-mb-md"
      data-ouia-component-id="OUIA-Generated-Breadcrumb-1"
      data-ouia-component-type="PF5/Breadcrumb"
      data-ouia-safe="true"
      data-testid="PfBreadcrumb"
    >
      <ol
        class="pf-v5-c-breadcrumb__list"
        role="list"
      >
        <li
          class="pf-v5-c-breadcrumb__item"
        >
          <a
            href="/insights/vulnerability/cves"
          >
            CVEs
          </a>
        </li>
        <li
          class="pf-v5-c-breadcrumb__item"
        >
          <span
            class="pf-v5-c-breadcrumb__item-divider"
          >
            <svg
              aria-hidden="true"
              class="pf-v5-svg"
              fill="currentColor"
              height="1em"
              role="img"
              viewBox="0 0 256 512"
              width="1em"
            >
              <path
                d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
              />
            </svg>
          </span>
        </li>
      </ol>
    </nav>
    <div
      class="pf-v5-l-split pf-m-gutter pf-m-wrap"
    >
      <div
        class="pf-v5-l-split__item"
      >
        <div
          class="pf-v5-l-flex pf-m-justify-content-space-between"
        >
          <div
            class=""
          >
            <h1
              class="pf-v5-c-title pf-m-2xl"
              data-ouia-component-id="OUIA-Generated-RHI/Header-true-2"
              data-ouia-component-type="RHI/Header"
              data-ouia-safe="true"
              widget-type="InsightsPageHeaderTitle"
            >
              CVEs
            </h1>
          </div>
        </div>
      </div>
      <div
        class="pf-v5-l-split__item"
      />
      <div
        class="pf-v5-l-split__item pf-m-fill"
      />
      <div
        class="pf-v5-l-split__item"
      >
        <button
          aria-expanded="false"
          class="pf-v5-c-menu-toggle dropdown-toggle"
          data-testid="dropdown-toggle"
          type="button"
        >
          <span
            class="pf-v5-c-menu-toggle__text"
          >
            Actions
          </span>
          <span
            class="pf-v5-c-menu-toggle__controls"
          >
            <span
              class="pf-v5-c-menu-toggle__toggle-icon"
            >
              <svg
                aria-hidden="true"
                class="pf-v5-svg"
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 320 512"
                width="1em"
              >
                <path
                  d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                />
              </svg>
            </span>
          </span>
        </button>
      </div>
    </div>
    <div
      class="pf-v5-l-grid pf-m-gutter"
    >
      <div
        class="pf-v5-l-grid__item pf-m-12-col-on-sm pf-m-8-col-on-md"
      >
        <div
          class="pf-v5-l-stack pf-m-gutter"
        >
          <div
            class="pf-v5-l-stack__item pf-v5-u-mt-sm"
          >
            <div
              class="pf-v5-l-flex"
            >
              <div
                class=""
              >
                Publish date:
              </div>
              <div
                class="pf-m-grow"
              >
                09 June 2020
              </div>
            </div>
          </div>
          <div
            class="pf-v5-l-stack__item"
          >
            <div
              class="pf-v5-c-content"
              style="text-align: justify;"
            >
              <p
                class=""
                data-ouia-component-id="OUIA-Generated-Text-1"
                data-ouia-component-type="PF5/Text"
                data-ouia-safe="true"
                data-pf-content="true"
              >
                A new domain bypass transient execution attack
              </p>
            </div>
          </div>
          <div
            class="pf-v5-l-stack__item pf-v5-u-mt-sm pf-v5-u-mb-md"
          >
            For most up-to-date information, 
            <a
              href="https://access.redhat.com/security/cve/CVE-2020-0543"
              rel="noopener noreferrer"
              target="_blank"
            >
              View in Red Hat CVE database
              <svg
                aria-hidden="true"
                class="pf-v5-svg pf-v5-u-ml-sm"
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 512 512"
                width="1em"
              >
                <path
                  d="M432,320H400a16,16,0,0,0-16,16V448H64V128H208a16,16,0,0,0,16-16V80a16,16,0,0,0-16-16H48A48,48,0,0,0,0,112V464a48,48,0,0,0,48,48H400a48,48,0,0,0,48-48V336A16,16,0,0,0,432,320ZM488,0h-128c-21.37,0-32.05,25.91-17,41l35.73,35.73L135,320.37a24,24,0,0,0,0,34L157.67,377a24,24,0,0,0,34,0L435.28,133.32,471,169c15,15,41,4.5,41-17V24A24,24,0,0,0,488,0Z"
                />
              </svg>
            </a>
          </div>
        </div>
      </div>
      <div
        class="pf-v5-l-grid__item pf-m-12-col-on-sm pf-m-4-col-on-md pf-v5-u-mt-sm pf-v5-u-ml-sm"
      >
        <div
          class="pf-v5-l-stack pf-m-gutter"
        >
          <div
            class="pf-v5-l-stack__item"
          >
            <div
              class="pf-v5-l-split pf-m-gutter"
            >
              <div
                class="pf-v5-l-split__item"
              >
                <div
                  style="display: contents;"
                >
                  <div
                    class="pf-v5-l-stack popover-content"
                  >
                    <div
                      class="pf-v5-l-stack__item"
                    >
                      <span
                        class="vuln-label "
                        style="font-size: medium; display: block;"
                      >
                        Business risk
                      </span>
                    </div>
                    <div
                      class="pf-v5-l-stack__item"
                    >
                      <a>
                        High
                      </a>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="pf-v5-l-split__item pf-v5-u-ml-lg"
              >
                <div
                  style="display: contents;"
                >
                  <div
                    class="pf-v5-l-stack popover-content"
                  >
                    <div
                      class="pf-v5-l-stack__item"
                    >
                      <span
                        class="vuln-label "
                        style="font-size: medium; display: block;"
                      >
                        Status
                      </span>
                    </div>
                    <div
                      class="pf-v5-l-stack__item"
                    >
                      <a>
                        <span>
                          <span
                            class="pf-v5-c-icon"
                          >
                            <span
                              class="pf-v5-c-icon__content"
                            >
                              <svg
                                aria-hidden="true"
                                class="pf-v5-svg pf-v5-u-mr-xs"
                                color="var(--pf-v5-global--primary-color--100)"
                                fill="currentColor"
                                height="1em"
                                role="img"
                                viewBox="0 0 576 512"
                                width="1em"
                              >
                                <path
                                  d="M569.517 440.013C587.975 472.007 564.806 512 527.94 512H48.054c-36.937 0-59.999-40.055-41.577-71.987L246.423 23.985c18.467-32.009 64.72-31.951 83.154 0l239.94 416.028zM288 354c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
                                />
                              </svg>
                            </span>
                          </span>
                           In review
                        </span>
                      </a>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="pf-v5-l-stack__item"
          >
            <span
              class="vuln-label pf-v5-u-mb-xs"
              style="font-size: medium; display: block;"
            >
              Severity
            </span>
            <span
              id="severity-shield"
              style="color: black;"
            >
              <span>
                <div
                  style="display: contents;"
                >
                  <span>
                    <span
                      class="pf-v5-c-icon pf-m-md"
                    >
                      <span
                        class="pf-v5-c-icon__content"
                      >
                        <svg
                          aria-hidden="false"
                          aria-label="Unknown"
                          class="pf-v5-svg"
                          color="#737679"
                          fill="currentColor"
                          height="1em"
                          role="img"
                          viewBox="0 0 384 512"
                          width="1em"
                        >
                          <path
                            d="M202.021 0C122.202 0 70.503 32.703 29.914 91.026c-7.363 10.58-5.093 25.086 5.178 32.874l43.138 32.709c10.373 7.865 25.132 6.026 33.253-4.148 25.049-31.381 43.63-49.449 82.757-49.449 30.764 0 68.816 19.799 68.816 49.631 0 22.552-18.617 34.134-48.993 51.164-35.423 19.86-82.299 44.576-82.299 106.405V320c0 13.255 10.745 24 24 24h72.471c13.255 0 24-10.745 24-24v-5.773c0-42.86 125.268-44.645 125.268-160.627C377.504 66.256 286.902 0 202.021 0zM192 373.459c-38.196 0-69.271 31.075-69.271 69.271 0 38.195 31.075 69.27 69.271 69.27s69.271-31.075 69.271-69.271-31.075-69.27-69.271-69.27z"
                          />
                        </svg>
                      </span>
                    </span>
                     Unknown
                  </span>
                </div>
              </span>
            </span>
          </div>
          <div
            class="pf-v5-l-stack__item"
          >
            <div
              class="pf-v5-c-content"
            >
              <div
                style="display: contents;"
              >
                <span
                  class="vuln-label pf-v5-u-mb-xs pointer"
                  style="font-size: medium; display: block;"
                >
                  CVSS 3.0  base score
                  <span
                    class="pf-v5-c-icon"
                  >
                    <span
                      class="pf-v5-c-icon__content"
                    >
                      <svg
                        aria-hidden="true"
                        class="pf-v5-svg pf-v5-u-ml-xs"
                        color="var(--pf-v5-global--secondary-color--100)"
                        fill="currentColor"
                        height="1em"
                        role="img"
                        viewBox="0 0 512 512"
                        width="1em"
                      >
                        <path
                          d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 448c-110.532 0-200-89.431-200-200 0-110.495 89.472-200 200-200 110.491 0 200 89.471 200 200 0 110.53-89.431 200-200 200zm107.244-255.2c0 67.052-72.421 68.084-72.421 92.863V300c0 6.627-5.373 12-12 12h-45.647c-6.627 0-12-5.373-12-12v-8.659c0-35.745 27.1-50.034 47.579-61.516 17.561-9.845 28.324-16.541 28.324-29.579 0-17.246-21.999-28.693-39.784-28.693-23.189 0-33.894 10.977-48.942 29.969-4.057 5.12-11.46 6.071-16.666 2.124l-27.824-21.098c-5.107-3.872-6.251-11.066-2.644-16.363C184.846 131.491 214.94 112 261.794 112c49.071 0 101.45 38.304 101.45 88.8zM298 368c0 23.159-18.841 42-42 42s-42-18.841-42-42 18.841-42 42-42 42 18.841 42 42z"
                        />
                      </svg>
                    </span>
                  </span>
                </span>
              </div>
              <span
                class="pf-v5-u-mr-md"
              >
                6.5
              </span>
              <span
                id="cvss-vector-content"
              >
                Vector: AV:L/AC:L/PR:L/UI:N/S:C/C:H/I:N/A:N
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>
  <section
    class="pf-v5-l-page__main-section pf-v5-c-page__main-section"
  >
    <div
      class="pf-v5-l-stack pf-m-gutter"
    >
      <div
        class="pf-v5-l-stack__item"
      >
        <div
          aria-label="Table title skeleton"
          class="pf-v5-c-skeleton"
          style="--pf-v5-c-skeleton--Width: 25%;"
        >
          <span
            class="pf-v5-screen-reader"
          >
            Loading contents
          </span>
        </div>
      </div>
      <div
        class="pf-v5-l-stack__item"
      >
        <div
          class="pf-v5-l-bullseye"
        >
          <svg
            aria-label="Contents"
            aria-valuetext="Loading..."
            class="pf-v5-c-spinner pf-m-lg"
            role="progressbar"
            viewBox="0 0 100 100"
          >
            <circle
              class="pf-v5-c-spinner__path"
              cx="50"
              cy="50"
              fill="none"
              r="45"
            />
          </svg>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
